core/src/main/java/hudson/util/MemoryReductionUtil.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.util;
+
+import hudson.Util;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Utilities to reduce memory footprint
+ * @author Sam Van Oort
+ */
+public class MemoryReductionUtil {
+    /** Returns the capacity we need to allocate for a HashMap so it will hold all elements without needing to resize. */
+    public static int preallocatedHashmapCapacity(int elementsToHold) {
+        if (elementsToHold <= 0) {
+            return 0;
+        } else if (elementsToHold < 3) {
+            return elementsToHold+1;
+        } else {
+            return elementsToHold+elementsToHold/3; // Default load factor is 0.75, so we want to fill that much.
```

Comment:
For numbers not divisible by 3, this goes slightly above 75%. Does that matter?

Reply:
@daniel-beck It may trigger a rehash but shouldn't be the end of the world.  Generally the HashMap uses power-of-two sizing anyway.